File: ios/SignaturePadType.swift
//
//  SignaturePadType.swift
//  ReactNativeModuleDemo
//
//  Created by Chintan Bawa on 25/05/24.
//

import UIKit

class SignaturePadType: UIView {
    private var signatureText = ""
    private var currentFontSize: CGFloat = 40.0
    private var currentFontSizeHalf: CGFloat {
        return currentFontSize / 2
    }
    private var canvasTopMargin: CGFloat = 0.0
    private var canvasMid: Int = 0
    private var attributes : [NSAttributedString.Key: Any]?
    
    // Views
    private var textPaint: NSMutableParagraphStyle!
    private var staticLayout: NSAttributedString?
    
    
    // Listeners
    private var sliderListener: SliderListener?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        self.backgroundColor = UIColor.white
        initTextPaint()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    func setSliderListener(listener: SignatureView) {
        sliderListener = listener
    }
    
    // Functions
    private func initTextPaint() {
        textPaint = NSMutableParagraphStyle()
        textPaint.alignment = .center
    }
    
    func setText(_ text: String) {
        signatureText = text
        
        if frame.width <= text.size(withAttributes: attributes).width {
            sliderListener?.onSliderProgressChange(value: Float(5))
            return
        }
        
        staticLayout = NSAttributedString(string: signatureText, attributes: attributes)
        
        let textHeight = staticLayout!.size().height
        canvasTopMargin = CGFloat(canvasMid) - (textHeight / 2)
        setNeedsDisplay()
    }
    
    func setSignatureFontSize(_ signatureFontSize: Float) {
        DispatchQueue.main.async {
            self.attributes?[.font] = UIFont(name: "Wood Dragon", size: CGFloat(signatureFontSize)) ?? UIFont(name: "Helvetica", size: CGFloat(signatureFontSize))
            self.setText(self.signatureText)
        }
    }
    
    // Prop functions
    func setStrokeColor(_ strokeColor: UIColor) {
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.5)  {
            self.attributes?[.foregroundColor] = strokeColor
            self.setText(self.signatureText)
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        // determine the width
        let width: CGFloat =  frame.width
        // determine the height
        let height: CGFloat = frame.height
        
        self.canvasMid = Int(height / 2)
        let maxFontSize = height
        let minFontSize = round(maxFontSize / 2)
        self.currentFontSize = maxFontSize
        attributes = [
            .font: UIFont(name: "Wood Dragon", size: currentFontSize) ?? UIFont(name: "Helvetica", size: CGFloat(currentFontSize))!,
            .paragraphStyle: textPaint!,
            .foregroundColor: UIColor.black
        ]
        
        sliderListener?.setSliderRange(min: Float(minFontSize), max: Float(maxFontSize))
        
        // Required call: set width and height
        frame.size = CGSize(width: width, height: height)
    }
    
    override func draw(_ rect: CGRect) {
        super.draw(rect)
        guard let staticLayout = staticLayout else { return }
        
        let context = UIGraphicsGetCurrentContext()
        context?.saveGState()
        context?.translateBy(x: 0, y: canvasTopMargin)
        staticLayout.draw(in: rect)
        context?.restoreGState()
    }
}

File: android/src/main/java/com/signatureview/SignaturePadType.kt
package com.signatureview

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Typeface
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import android.view.View
import com.signatureview.interfaces.SeekBarListener


class SignaturePadType : View {
  private var signatureText = ""
  private var currentFontSize = 150f
  private var currentFontSizeHalf = currentFontSize / 2
  private var canvasTopMargin = 0f
  private var canvasMid = 0

  //views
  private lateinit var textPaint: TextPaint
  private var staticLayout: StaticLayout? = null

  //listeners
  private var seekBarListener: SeekBarListener


  constructor(signatureView: SignatureView) : super(signatureView.context) {
    seekBarListener = signatureView
    initTextPaint()
  }

  // Functions
  private fun initTextPaint() {
    textPaint = TextPaint()
    textPaint.isAntiAlias = true
    textPaint.textSize = currentFontSize
    textPaint.color = Color.parseColor("#000000")
    textPaint.typeface = Typeface.createFromAsset(context.assets, "fonts/signature.ttf");
  }

  fun setText(text: String) {
    signatureText = text

    if (signatureText.length > 1 && width < textPaint.measureText(
        signatureText
      )
    ) {
      seekBarListener.onSeekBarProgressChange(10)
      return
    }

    val builder = StaticLayout.Builder.obtain(
      signatureText, 0, signatureText.length, textPaint, width
    ).setAlignment(Layout.Alignment.ALIGN_CENTER).setLineSpacing(1f, 0f).setIncludePad(false)

    staticLayout = builder.build()

    canvasTopMargin = canvasMid - (staticLayout!!.lineCount * currentFontSizeHalf)
    invalidate()
  }

  fun setSignatureFontSize(signatureFontSize: Int) {
    if (staticLayout != null) {
      this.currentFontSize = signatureFontSize.toFloat()
      this.currentFontSizeHalf = currentFontSize / 2
      textPaint.textSize = currentFontSize
      setText(signatureText)
    }
  }

  // Prop functions
  fun setStrokeColor(strokeColor: Int) {
    post(Runnable {
      textPaint.color = strokeColor
      setText(signatureText)
    })
  }

  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    val width = MeasureSpec.getSize(widthMeasureSpec)
    val height = MeasureSpec.getSize(heightMeasureSpec)
    this.canvasMid = height / 2
    val maxFontSize = height
    val minFontSize = Math.round((maxFontSize / 2).toDouble())
    this.currentFontSize = maxFontSize.toFloat()
    this.currentFontSizeHalf = currentFontSize / 2
    textPaint.textSize = currentFontSize
    seekBarListener.setSeekBarRange(minFontSize.toInt(), maxFontSize)

    // Required call: set width and height
    setMeasuredDimension(width, height)
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    if (staticLayout == null) return

    canvas.save()
    canvas.translate(0f, canvasTopMargin)
    staticLayout!!.draw(canvas)
    canvas.restore()
  }
}
